### ISSUE TITLE
prepDE.py AttributeError

### ISSUE THREAD

**Issue by jstrohm** [gpertea/stringtie#83]
Hello, 

I’ve been trying to work through this error for a few days now. I've opened and closed a few issues as I found mistakes that I made. However, I have now run out of ideas. 

I have run the script successfully using your example data and my set of commands. Which means that the issue is with one of my input files. I suspect it's the reference annotation. I have used gffread -E to scan for potential issues and none were found. 

Once again, the error I am getting is: 

./prepDE.py -i /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown_test/ -g /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown/test_gene_count_matrix.csv -t /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown/test_transcript_count_matrix.csv
Traceback (most recent call last):
  File "./prepDE.py", line 71, in <module>
    g_id=RE_GENE_ID.search(v[len(v)-1]).group(1)
AttributeError: 'NoneType' object has no attribute 'group'

I have attached my reference genome annotation.gtf and a ballgown.gtf file that I generated. Below are the commands that I am using. Thank you!

./gffread -E /media/genetics/Data1/Jeff_workstation_1/subread-1.5.0-p1/GCF_000233375.1_ICSASG_v2_genomic.gff -o /media/genetics/Data1/Jeff_workstation_1/subread-1.5.0-p1/GCF_000233375.1_ICSASG_v2_genomic_filter.gff

./stringtie /media/genetics/Data1/Jeff_workstation_1/hisat2-2.0.5/mapped_reads/BAM_sorted/3-B5389_S1_AMY2P_NGSLib1d_mRNA.sorted.bam -v -o /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf -p 8 -G /media/genetics/Data1/Jeff_workstation_1/subread-1.5.0-p1/GCF_000233375.1_ICSASG_v2_genomic_filter.gtf -l 3-B5389_S1_AMY2P_NGSLib1d_mRNA

./stringtie --merge /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/2-B5399_S1_mRNA_ART25_NGSLib2.gtf /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/3-B5389_S1_1a_mRNA.gtf /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf -G /media/genetics/Data1/Jeff_workstation_1/subread-1.5.0-p1/GCF_000233375.1_ICSASG_v2_genomic_filter.gtf -o /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/test_merge.gtf

./stringtie -e -B -p 8 -G /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/test/test_merge.gtf -o /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown_test/3-B5389_S1_AMY2P_NGSLib1d_mRNA/3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf /media/genetics/Data1/Jeff_workstation_1/hisat2-2.0.5/mapped_reads/BAM_sorted/3-B5389_S1_AMY2P_NGSLib1d_mRNA.sorted.bam

python ./prepDE.py -i /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown_test/ -g /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown/test_gene_count_matrix.csv -t /media/genetics/Data1/Jeff_workstation_1/stringtie-1.3.1c.Linux_x86_64/ballgown/test_transcript_count_matrix.csv

[ballgown_example.tar.gz](https://github.com/gpertea/stringtie/files/617080/ballgown_example.tar.gz)

[salmo_salar_ICSASG_v2_genomic_filter.gtf.tar.gz](https://github.com/gpertea/stringtie/files/617082/salmo_salar_ICSASG_v2_genomic_filter.gtf.tar.gz)

**Comment by jstrohm** [comment]
Looking at your example file chrX.gtf it has the naming structure:

chrX	hg38_refGene	exon	50288780	50288887	0	+	0	gene_id "NM_033031"; transcript_id "NM_033031";  gene_name "CCNB3"; product "G2/mitotic-specific cyclin-B3 isoform 3";

But when I use gffread to convert the .gff version of my annotation to .gtf, it is in this format:

NC_027300.1	Gnomon	exon	162535	162651	0	-	0	transcript_id "rna1"; gene_id "gene1"; gene_name "LOC106607996";

Could this be the issue?

**Comment by gpertea** [comment]
The naming convention should not matter. Just to be sure, could you confirm that it's python 2 being run there (not python 3). ```python --version``` should show this.

All those full paths and file names are kind of hard to follow (for human eyes), I hope you got them all right in your scripts there..
I only see the ```stringtie -e -B``` command for the 3-B5389_S1_1a_mRNA sample and I see you are running this for 3 samples -- could you please attach (or otherwise make available to me) the other 2 output .gtf files you got under the **ballgown_test**/ directories when you ran ```stringtie -e -B``` for each sample? I see you attached 2-B5399_S1_mRNA_ART25_NGSLib2.gtf which seems to be one of the outputs, so I guess I now just need 3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf  and presumably 3-B5389_S1_1a_mRNA.gtf (if you followed the same hard-to-follow naming convention) in order to reproduce the error here.

**Comment by jstrohm** [comment]
Sorry my sample names are so clunky... I will choose something easier for troubleshooting going forward if providing the other output.gtf files doesn’t provide a clear solution. Thanks for taking a look for me!

Yes I am using Python 2.7.12

[3-B5389_S1_1a_mRNA.gtf.tar.gz](https://github.com/gpertea/stringtie/files/617535/3-B5389_S1_1a_mRNA.gtf.tar.gz)

[3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf.tar.gz](https://github.com/gpertea/stringtie/files/617537/3-B5389_S1_AMY2P_NGSLib1d_mRNA.gtf.tar.gz)

**Comment by gpertea** [comment]
Thanks for the files -- aha, I see what the problem is, some of the reference transcripts lack the *gene_id* attribute while the prepDE.py script assumes that it's always there.. But it's quite easy to fix this (at least prevent the crash). Unfortunately it's late now, my Python skills are rusty and I need some sleep -- but tomorrow morning I'm going to provide a quick-and-dirty fix for this.

**Comment by gpertea** [comment]
I updated prepDE.py script on the main website (added to this git repo as well), please downloaded it from here:

http://ccb.jhu.edu/software/stringtie/dl/prepDE.py

It includes the fixes for regular expression parsing of gene_id (and I threw in a coverage parsing check as well).

**Comment by jstrohm** [comment]
Thank you for fixing this so quickly Geo!

As a quick side-question, what is the simplest way to include the gene "product" field i.e. full gene names in the output of prepDE.py? Could I use gffcompare in annotation mode for the input ballgown.gtf files?

**Comment by gpertea** [comment]
The simplest way usually depends on your scripting abilities ;)
The "product" and/or "name" info are not currently included in the CSV output of prepDE (due to the minimalist format requirements, but mostly because they might not be available) but it's not that hard to link those gene_id (or transcript_id) to gene names and product descriptions in some post-processing stage. This really depends on the availability of annotation information for all those transcripts/genes.
